fml/message_loop.cc

```diff
@@ -15,64 +15,96 @@
 namespace fml {
 
 FML_THREAD_LOCAL ThreadLocal tls_message_loop([](intptr_t value) {
-  delete reinterpret_cast<MessageLoop*>(value);
+  delete reinterpret_cast<RefPtr<MessageLoop>*>(value);
 });
 
 MessageLoop& MessageLoop::GetCurrent() {
-  auto loop = reinterpret_cast<MessageLoop*>(tls_message_loop.Get());
+  auto loop = reinterpret_cast<RefPtr<MessageLoop>*>(tls_message_loop.Get());
   FML_CHECK(loop != nullptr)
       << "MessageLoop::EnsureInitializedForCurrentThread was not called on "
          "this thread prior to message loop use.";
-  return *loop;
+  return *loop->get();
 }
 
 void MessageLoop::EnsureInitializedForCurrentThread() {
   if (tls_message_loop.Get() != 0) {
     // Already initialized.
     return;
   }
-  tls_message_loop.Set(reinterpret_cast<intptr_t>(new MessageLoop()));
+  auto message_loop_ptr = new RefPtr<MessageLoop>(nullptr);
+  *message_loop_ptr = MakeRefCounted<MessageLoop>();
+  tls_message_loop.Set(reinterpret_cast<intptr_t>(message_loop_ptr));
 }
 
 bool MessageLoop::IsInitializedForCurrentThread() {
   return tls_message_loop.Get() != 0;
 }
 
-MessageLoop::MessageLoop()
-    : loop_(MessageLoopImpl::Create()),
-      task_runner_(fml::MakeRefCounted<fml::TaskRunner>(loop_)) {
-  FML_CHECK(loop_);
-  FML_CHECK(task_runner_);
+MessageLoop::MessageLoop() {
+  PushMessageLoop();
 }
 
 MessageLoop::~MessageLoop() = default;
 
-void MessageLoop::Run() {
-  loop_->DoRun();
+void MessageLoop::Run(fml::closure on_done) {
```

Comment:
Spoke to Jason about this and there is definitely a confusion about the new API. The `on_done` callback is a place for callers to post task onto the new message loop. This is because posting tasks onto the message loop before the run call will post the same onto the wrong loop activation, and, posting them after the run call is not possible because the call is blocking.

The consensus seems to be that modifying the semantics of the run call is potentially confusing. Instead, a `RunNested` call that takes a closure of tasks to run on the nested message loop (and then terminate the same) would be a clear indication to callers that the new run semantics hold.